mlflow/models/evaluation/default_evaluator.py

```diff
@@ -1049,3 +1042,44 @@ def evaluate(
                     return self._evaluate_regressor()
                 else:
                     raise ValueError(f"Unsupported model type {model_type}")
+
+    @property
+    def X(self):
+        """
+        The features (`X`) portion of the dataset, guarded against accidental mutations.
+        """
+        return DefaultEvaluator._MutationGuardedData(
+            pd.DataFrame(self.dataset.features_data, columns=self.dataset.feature_names)
+        )
+
+    @property
+    def y(self):
+        """
+        The labels (`y`) portion of the dataset, guarded against accidental mutations.
+        """
+        return DefaultEvaluator._MutationGuardedData(self.dataset.labels_data)
+
+    class _MutationGuardedData:
+        """
+        Wrapper around a data object that only provides copies of the data object, guarding against
+        accidental reuse of the data object after it is mutated, e.g. during inference with
+        scikit-learn models.
+        """
+
+        def __init__(self, data):
+            """
+            :param data: A data object, such as a Pandas DataFrame, numPy array, or list.
+            """
+            self._data = data
+
+        def copy_in_case_of_mutation(self):
```

Comment:
Prefer `copy_in_case_of_mutation()` over `copy()` so that, when reading the code, it becomes clear why we're obtaining a new copy of the dataset.

Reply:
It might be my English, but the name `copy_in_case_of_mutation` made me think it may not perform a copy operation if there's no mutation on the DataFrame. But the way it's implemented is actually always-copy. Shall we rename it to avoid such confusion? e.g. `copy_to_avoid_mutation`(). IMO simple `copy()` also reads pretty clear to me what it's trying to achieve, or something like `pass_by_copy()`, `get_copy()`

Reply:
Agreed! I've changed this to `copy_in_case_of_mutation()`, which more clearly gets at the motivation. Naively, a reader might wonder why we're calling `copy()`; I think `copy_in_case_of_mutation()` makes things slightly clearer.